### PR TITLE
Performance improvement for processing

### PR DESCRIPTION
**Pull request by sritterginkgo**

### CODE REVIEW

**Comment by sritterginkgo** [comment]
A little new to adding to public, let me know.

I didn't add new tests for this, I ended up rewriting the logic and re-using some functionally that exists. I ran your existing tests on writer.py which passed. Looking over those tests they seem to cover logic that I would've tested for. Let me know!

File: src/litdata/streaming/writer.py [hunk 1]
```diff
@@ -289,23 +293,47 @@ def add_item(self, index: int, items: Any) -> Optional[str]:
             bytes=len(data),
             dim=dim,
         )
-
-        if not self._should_write():
+        if self._min_index is None:
+            indexes = list(self._serialized_items.keys())
+            self._min_index = index = indexes[0] if len(indexes) == 1 else min(*indexes)
```
Comment (tchaton) [inline src/litdata/streaming/writer.py:298]:
```suggestion
            self._min_index = indexes[0] if len(indexes) == 1 else min(*indexes)
```

File: src/litdata/streaming/writer.py [hunk 2]
```diff
@@ -289,23 +293,47 @@ def add_item(self, index: int, items: Any) -> Optional[str]:
...
 
     def _should_write(self) -> bool:
         # TODO: Misleading method name, it modifies `self._min_index` and `self._max_index`!
         if not self._serialized_items:
             return False
-        indexes = list(self._serialized_items.keys())
-        self._min_index = index = indexes[0] if len(indexes) == 1 else min(*indexes)
-        num_bytes = 0
-        num_items = 0
+        num_bytes = self._per_sample_num_bytes
```
Comment (tchaton) [inline src/litdata/streaming/writer.py:334]:
```suggestion
        # We have already validated min_index to max_index. Resetting the num_bytes and  num_items.
        num_bytes = self._per_sample_num_bytes
```

**Review by tchaton — APPROVED** [review]
Looks great !

**Comment by tchaton** [comment]
@sritterginkgo Would you like to join our Slack or Discord ? I am creating a channel about litdata.

**Comment by sritterginkgo** [comment]
Sure!

**Comment by tchaton** [comment]
@sritterginkgo  Congratulations on your first PR. Keep them coming :) Reach out if you need anything on Discord.

**Comment by tchaton** [comment]
@sritterginkgo what's your username on Discord ? I would like to invite to the litdata channel.

**Pull request merged**